persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/client/ClientDAOImpl.kt

```diff
@@ -32,6 +32,14 @@ internal class ClientDAOImpl internal constructor(
         }
     }
 
+    override suspend fun insertClientsAndRemoveRedundant(qualifiedID: QualifiedIDEntity, clients: List<Client>) =
+        clientsQueries.transaction {
+            clients.forEach { client ->
+                clientsQueries.insertClient(client.userId, client.id, client.deviceType)
+            }
+            clientsQueries.deleteClientsOfUserExcept(qualifiedID, clients.map { it.id })
+        }
```

Comment:
Considering this is done in a loop, using [Grouped Statements](https://cashapp.github.io/sqldelight/multiplatform_sqlite/grouping_statements/) is significantly better than just a transaction, performance-wise.